### DISCUSSION TITLE
#16 - Technical Review 🛠👀

### DISCUSSION THREAD

**Discussion by Annalytic-programming** [CUMGroup/AMOGUS#135] [Blog]
# Howdy AMOGUS friends, 🤠

![](https://media.tenor.com/ysX0T3VItMMAAAAC/howdy-spongebob.gif)

last week we learned a lot about technical reviews. Of course, we immediately put what we learned into practice and held our own technical review as part of a formal meeting and we don't want to withhold the result of it from you. So sit back, grab a coffee ☕ or energy drink and let the review begin.

## The Preparation

Each of us had already thought about what should be addressed in the meeting, in advance. That's why we agreed relatively quickly on which components of AMOGUS we were talking about. 
The components were:
- the game component (frontend): game view and  logic
- the exercise service (backend): get exercises of specific category and validate answer
- the mental exercise factory (backend): generating random mental exercises

These three are more or less the heart ❤ of AMOGUS, which is why the choice fell on them.

## The Review itself

We decided to look at each component separately and therefore use different methods to review them.
However, what we did with all of them was that we did a short walkthrough 👀 to get an overview of the code again.
The following methodologies seemed to be the most suitable for the components:

| Component | Methodology |
| - | - |
| game component | formal inspection 👀 |
| exercise service | benchmarks 📊 |
| mental exercise factory | benchmarks 📊 |

The review also showed us some weaknesses that need to be reworked towards the end. Here is a brief overview of which actions we have decided on which components:

| Component | Action |
| - | - |
| game component | code refactoring 🛠 |
|  | encode JSON session object to base64 🔐 |
| exercise service | benchmarks 📊 |
| mental exercise factory | benchmarks 📊 |
|  | code refactoring 🛠 |

## What we have learned 

We learned a lot from this review. Finally, here are our lessons learned:
- security is not that easy to achieve
- code can easily get very complicated to understand (clean code matters)
- external libraries can help to simplify the code
- add website source links to comments if needed
- always taking the hobbits to isengard 💍

We may have picked up one of these lessons elsewhere but we cannot remember. 🤔

As you already know, there are also the AMOGUS docs. Of course, we have also uploaded our artifact for the review [here](https://cumgroup.github.io/AMOGUS/Developer%20Docs/8.%20Review/technical_review/), if you need further information.
You can also check the results of the benchmarks [here](https://cumgroup.github.io/AMOGUS/Developer%20Docs/5.%20Metrics/Benchmarks/).
As you can see above, we decided to base64 encode the session object.
This is mainly due to the fact that the answers of our questions can be viewed via the browser's network tab.
To make it much more difficult for users to cheat, we now encode the object. You can view the corresponding pull request [here](https://github.com/CUMGroup/AMOGUS/pull/133).
And one more [link](https://github.com/CUMGroup/AMOGUS/pull/134) for you if you are interested in our refactoring of the mental exercise factory.

We hope you had fun reading it.
Think of your [AMOGUS](https://amogus.alexmiha.de/) streak. See you next week 👋

Yours truly,
CUMGroup

**Comment by Luca-Hackl** [comment]
Ahohoy,
Thanks for your interesting insights into your retrospective. Some more data would have been nice but other than this small critique point we have nothing to meckern :)

Greetings
RALF

**Reply by AlexMi-Ha** [reply]
Hi RALF,
Thanks for your feedback. What more data would you be missing here?
We included all relevant documents and actions. Feel free to ask for specifics!

Yours truly,
CUMGroup

**Comment by vans1646** [comment]
Hello there,
it's really nice that you've decided to not only review one of your components but all the main parts. 
I think this way you can really make sure to cover everything. 

Especially it's really good that you considered a code refactoring should be done for some parts. Improving old stuff is always a good way to start, just make sure it gets better and your not breaking anything (never change a running system and so on). But I'm pretty confident that you will do great and your project will get even better.

Looking forward to reading more from you.

Cheers,
Capyclue

**Reply by AlexMi-Ha** [reply]
Hi Capyclue,
nice to hear from you! ☺

We thought that inspecting all core elements was the right way to go, as not every member of the team has seen everything in those.
Now we are all up to date with the components, which helps discussing specifics easier in the future!

The old implementation of the MentalExerciseFactory was not bad, but the mathematics were a bit _all over the place_. It was very fun checking the equivalence of the xp-calculation after the refactor 💀^^

Yours truly,
CUMGroup

**Comment by 2000eBe** [comment]
Hey, 
your blogposts suggest you're very organised and it does seem to be the case :) 
Its nice to see some benchmarks in your case and that you openly show them. Transparency is always a good trait!
I also took a look into your merged refactoring pull requests (transparency again - hurrah!). Most of the time refactoring is seen as a tedious and annoying task, but hopefully you'll swiftly notice some positive feedback from your own team handling the code!

(also... Legolas would be proud of you for taking the hobbits to Isengard)

Greetings
VIGAD (B6)

**Reply by AlexMi-Ha** [reply]
Hi VIGAD,
thanks for the lovely comment!^^

Nice to hear you've looked at our documentation and prs. Transparency is important for us (and we have _nothing_ to hide 👀)
The refactor definitely helped with the readability of our exercise calculation process.

(he is 💍)

Yours truly,
CUMGroup

**Comment by MarcGriese** [comment]
Great job! 
We liked how you organized the review process and took action on identified weaknesses. Sharing lessons learned was valuable for us too. 
For improvement, consider providing more specific details, including visuals, and sharing insights behind your decisions. Overall, a clear and informative blog post. Keep up the good work!

Cheers, 
DHBWorkout

**Reply by AlexMi-Ha** [reply]
Hello there DHBWorkout,
with our decisions we tried to make a more solid foundation for the whole project.
Our `ExerciseService` and the `MentalExerciseFactory` are part of the heart of AMOGUS and used really heavily. So having insights on the Performance with Benchmarks was important to us. To increase maintainability of the `factory` a refactor of the maths involved was definitely overdue!
Furthermore we tried to look at Security - like ensuring that players cannot cheat easily. The main point here was that the answers of the questions were directly written (in plain text) in the network request and could be easily viewed. This is why we decided to take action to encode it for the network request and decode it later in the game.

I hope this provided some more insights! 😊

Yours truly,
CUMGroup
